### PR TITLE
Auto-minimization

### PR DESCRIPTION
**Pull request by SolalPirelli**
opening this already just so we can discuss it if needed

### CODE REVIEW

File: src/it/scala/inox/solvers/OptimizerSuite.scala [hunk 1]
```diff
@@ -146,4 +146,26 @@ class OptimizerSuite extends SolvingTestSuite with DatastructureUtils {
...
+*/
+  test("automated minimization of n times n") { implicit ctx =>
+    val x = Variable.fresh("x", Int32Type())
+    val prop = GreaterEquals(Times(x, x), Int32Literal(10))
+
+    val factory = SolverFactory.optimizer(program, ctx)
+    val optimizer = factory.getNewSolver()
+    try {
+      optimizer.assertCnstr(Not(prop))
+      optimizer.check(Model) match {
```
Comment (SolalPirelli) [inline src/it/scala/inox/solvers/OptimizerSuite.scala]: look ma, no explicit `minimize`!

File: src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+     * (The concept of "smaller" is not well-defined either, e.g., one could reasonably argue that P(a: BV16, b: BV16) is either "smaller" or "larger" than Q(c: BV64)
+     *  due to the size in bits vs number of fields; we go with the number of fields)
+     */
+    def adtSizers(ctors: Seq[TypedADTConstructor]): Seq[Expr] = {
+      /** Gets the 'zero' literals for the given field */
+      def getZeroLiterals(field: ValDef): Seq[Expr] = field.tpe match {
+        case BVType(signed, size) => Seq(BVLiteral(signed, 0, size))
+        case adt @ ADTType(_, _) =>
+          adt.lookupSort match {
+            case Some(tsort) => tsort.constructors.flatMap(_.fields).distinct.flatMap(getZeroLiterals)
```
Comment (mbovel) [inline src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala]: We get an infinite recursion here if the ADT is recursive, for example:

```smt2
(declare-datatypes () (
    (MyList
        (MyNil)
        (MyCons (x Int) (xs MyList))
    )
))


(define-fun-rec myMin ((t MyList)) Int
    (match t 
        (case MyNil 1000)
        (case (MyCons x xs)
            (let ((xsMin (myMin xs)))
            (ite (<= x xsMin) x xsMin))
        )
    )
)

(define-fun-rec myMax ((t MyList)) Int
    (match t 
        (case MyNil (- 1))
        (case (MyCons x xs)
            (let ((xsMax (myMax xs)))
            (ite (>= x xsMax) x xsMax))
        )
    )
)

(declare-fun aList () MyList)

(assert (>= (myMin aList) 0))
(assert (<= (myMin aList) 999))
(assert (not (= (myMin aList) (myMax aList))))


(check-sat)
```

Let's try to just stop the recursion in such cases (do nothing for nested objects). I think that Inox creates free variables for them when unfolding, so they will be handled by the outmost loop anyway.

(This is TIP code, runnable with inox using `sbt compile` and then `inox -v --solvers=smt-z3 lists.tip`.)

File: src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+trait Z3Minimizer extends Z3Optimizer {
+  import program._
+  import program.trees._
+  import program.symbols._
+  import exprOps.variablesOf
+
+  private def sizersOf(e: Expr): Seq[Expr] = {
+    /**
+     * Gets the expressions that should be minimized for the given ADT constructors.
+     * This is not as simple as it first appears because of the split between the bitvector and integer worlds;
```
Comment (mbovel) [inline src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala:37]: I still get some errors when mixing bitvectors and integers, for example:

```scala
object ADT {
    sealed trait X
    final case class X1(a: Int) extends X
    final case class X2(a: BigInt, b: Int) extends X

    sealed trait Y
    final case class Y1(a: Z) extends Y
    final case class Y2(b: X) extends Y

    final case class Z(a: X)

    def getLong(x: X): Long = x match {
        case X1(a) => a
        case X2(_, b) => b
    }

    def getLong(y: Y): Long = y match {
        case Y1(a) => getLong(a.a)
        case Y2(b) => getLong(b)
    }

    def f(x: X, y: Y): Long = {
        getLong(x) - getLong(y)
    }.ensuring(_ <= -4)
}
```

Error:

```
Internal solver error in z3: line 24 column 144: Sorts (_ BitVec 32) and Int are incompatible
```

Generated query:

```smt2
; Options: -in -smt2
(set-option :produce-unsat-assumptions true)

(declare-fun start!4 () Bool)

(assert start!4)

(declare-fun lt!5 () (_ BitVec 64))

(minimize (ite (bvsge lt!5 #b0000000000000000000000000000000000000000000000000000000000000000) lt!5 (bvneg lt!5)))
(declare-datatypes () ((X!2 (X1!2 (a!62 (_ BitVec 32))) (X2!2 (a!63 Int) (b!39 (_ BitVec 32))))))

(declare-datatypes () ((Z!4 (Z!5 (a!64 X!2)))))

(declare-datatypes () ((Y!2 (Y1!2 (a!65 Z!4)) (Y2!2 (b!40 X!2)))))

(declare-fun y!22 () Y!2)

(get-info :version)

(minimize (ite ((_ is Y1!2) y!22) #b00000000000000000000000000000001 #b00000000000000000000000000000001))
(minimize (ite ((_ is Y1!2) y!22) #b00000000000000000000000000000001 #b00000000000000000000000000000000))
(minimize (ite ((_ is Y1!2) y!22) (ite ((_ is X1!2) (a!64 (a!65 y!22))) #b00000000000000000000000000000001 #b00000000000000000000000000000010) 0))
```

File: src/it/scala/inox/solvers/MinimizerSuite.scala [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        case SatWithModel(model) =>
+          model.vars.get(x.toVal).get match {
+            case Int32Literal(c) => assert(c == 0)
+          }
+        case _ =>
+          fail("Expected sat-with-model")
+      }
+    } finally {
+      factory.reclaim(optimizer)
+    }
```
Comment (mbovel) [inline src/it/scala/inox/solvers/MinimizerSuite.scala:40]: We'll probably want more tests here.

**Comment by mbovel** [comment]
Note: Z3 is generally not able to optimize recursive functions or quantified expressions.

See: https://github.com/Z3Prover/z3/issues/1382.

For example:

```smt2
(define-fun-rec sumToTen ((n Int)) Int
    (ite (>= n 4) 4 (+ (sumToTen (+ n 1)) n)))

(declare-fun k () Int)
(declare-fun l () Int)

(assert (= l (sumToTen k)))

(assert (>= k 0))
(assert (>= l 4))
(assert (<= l 7))
(minimize k) ; remove this line when running with Inox

(check-sat)
(get-model) ; remove this line when running with Inox
```

Here, Z3 and Inox will yield `k = 4` instead of `k = 3`. That makes sense as these models are generally unbounded. In these cases, Z3 and Inox will unfold functions incrementally and stop as soon as a valid model is found.

**Comment by mbovel** [comment]
Nice work! Cool to have the Minimizer modularized in its own class! And I think that having separate `minimize` commands for each field is the right thing to do. This will allow us to use the Z3 built-in ways to combine objectives ( https://theory.stanford.edu/~nikolaj/programmingz3.html#sec-multiple-objectives) if needed.

**Comment by mbovel** [comment]
I'd be up to move this PR to the Inox repo as soon as possible as we might gain from comments from Nicolas, Viktor and co.

File: src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala (hunk 2 above)
Comment (SolalPirelli) [reply src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala]: Oops, completely forgot about those, thanks. Fixed.

File: src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala (hunk 3 above)
Comment (SolalPirelli) [reply src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala:37]: This seems to have been fixed alongside the recursive fix

**Comment by SolalPirelli** [comment]
I've fixed the stack overflow (thanks!), I'll open a PR to the main repo now and also ask what kind of tests they'd like / if there's a shorter way to write these tests than the super-verbose Inox DSL.

**Comment by mbovel** [comment]
> a shorter way to write these tests than the super-verbose Inox DSL.

The TIP format looks like a good fit, doesn't it?

It seems easy to load files in this format using `Parser.parseScript` as in https://github.com/epfl-lara/inox/blob/master/src/it/scala/inox/tip/TipTestSuite.scala#L71.

File: src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala (hunk 2 above)
Comment (mbovel) [reply src/main/scala/inox/solvers/smtlib/optimization/Z3Minimizer.scala]: Cool, thanks!

**Pull request closed without merging**